[WIP][12.0] account_customer_wallet, pos_customer_wallet

carmenbianca opened the pull request:
## Description



## Odoo task (if applicable)

https://gestion.coopiteasy.be/web#id=4393&view_type=form&model=project.task&action=475&active_id=200

## Checklist before approval

- [x] Tests are present (or not needed).
- [x] Credits/copyright have been changed correctly.
- [ ] (If a new module) Moving this to OCA has been considered → Submit to OCA

---

- [x] Verify the name `account_customer_wallet` and `pos_customer_wallet` (and the user-facing terminology).
- [x] Create a global toggle for the module in General Settings
- [x] Create a global setting for the customer wallet account.journal
- [x] Don't allow the global custom wallet account.journal to be undefined if the global toggle is enabled.
- [x] On res.partner, create a field 'balance of customer wallet' which is calculated.
- [x] Display the above field in the 'Sales & Purchases' tab of partner.
- [ ] Add permissions to fields/views/whatever.
- [ ] Test for rounding errors.
- [x] Add a toggle 'is customer wallet' to account.journal.
- [x] When a client is selected in POS, show their balance. Condition: IF global module toggle is enabled
- [ ] [NICE TO HAVE] … AND the customer wallet payment method is selected.
- [x] POS sets partner's parent on `account.move.line`s. Decide whether this is desirable behaviour.
- [x] Update customer wallet balance in POS after a payment is made in the POS.
  - [x] Implement the logic in Python
  - [x] [NICE TO HAVE] Refresh customer data from DB when customer is selected in POS and POS is online. This keeps things synchronous between POS sessions.
- [x] Repair expectedfailure tests.
- [x] In POS: Make sure customer is selected when:
  - [x] selecting/using wallet payment method
  - [ ] [DON'T YET IMPLEMENT] adding to wallet balance
- [x] When a customer tries to make a payment with more money than they have in their balance, fail.
- [ ] [NICE TO HAVE] If the balance is <0, colour it RED.
- [ ] [NICE TO HAVE] Test performance considerations of making the balance a computed field. (Spec solution: Only trigger the calculation of the balances at the start of the session. (what if there are multiple sessions?) Condition: IF global module toggle is enabled AND a customer method payment method is available.)

carmenbianca committed:
[ADD] account_customer_wallet

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] account_customer_wallet: is_customer_wallet_journal field

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] account_customer_wallet: Global settings

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] account_customer_wallet: Make the global journal option mandatory if the global toggle is enabled

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] account_customer_wallet: Create customer_wallet_balance for res.partner

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] pos_customer_wallet

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] account_customer_wallet: Use account instead of journal

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] account_customer_wallet: Save customer wallet account on res.company

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] pos_customer_wallet: Display wallet balance on POS payment screen

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] pos_customer_wallet: Make global module toggle available in pos.config

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] account_customer_wallet: Demo account/journal data

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] pos_customer_wallet: Journal demo data

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] account_customer_wallet: Scaffolding + logic for balance computation

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] pos_customer_wallet: Mandate selected customer when using wallet payment method

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] pos_customer_wallet: Check sufficient balance before POS purchase

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] account_customer_wallet: Test what happens when the wallet account is un-configured

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca commented on account_customer_wallet/tests/test_balance.py:
```diff
@@ -101,3 +101,21 @@ def test_debit_and_credit(self):
+            "account_customer_wallet.customer_wallet_account_id",
+            None,
+        )
+        self.assertFalse(
+            self.env["ir.config_parameter"].get_param(
+                "account_customer_wallet.customer_wallet"
+            )
+        )
```
Not sure why the test fails here. It's like `_onchange_customer_wallet` in `res.config.settings` is never triggered. As a matter of fact, it isn't—I tested in the debugger.

robinkeunen replied:
`res.config.settings` is a TransientModel with helper functions, some of which use `set_param` or `get_param` to write in `ir.config_parameter`.

So here, no `res.config.settings` is created for the onchange to be triggered.

More over : the onchange functions are only triggered when **forms** are modified. In tests, you need to use the `with Form(some object):` to test that behavior.

All of this is probably obsolete if you set the required constraint on the form as I suggested earlier.

carmenbianca commented:
Problem:

1. Customer has a customer wallet balance of €70.
2. Customer buys something, using €20 of their wallet balance.
3. Customer leaves the shop and later comes back to purchase a €60 item.
4. At the POS payment screen, the customer's wallet balance _still displays €70_!

This is, I assume, because the `account.move.line`s aren't created/verified until after the POS session is closed and verified.

Solution:

1. Write some JavaScript to calculate the current balance: `partner.customer_wallet_balance - all payments made by the partner in the current session`. Problem: Doesn't work across sessions.
2. Create another field `pos_session_customer_wallet_balance` on partner that is equal to `partner.customer_wallet_balance - all payments made by the partner across all active POS sessions`. Problem: I'm not sure what `@api.depends` to use here. I think I'd have to create a computed Many2many field on res.partner that returns all active POS sessions, and I can then do something like `@api.depends("pos_session_ids, pos_session_ids.payment_lines")` (or whatever the actual field would be).

@robinkeunen @legalsylvain What are your thoughts?

robinkeunen commented on account_customer_wallet/models/res_config_settings.py:
```diff
@@ -0,0 +1,43 @@
+class ResConfigSettings(models.TransientModel):
+    _inherit = "res.config.settings"
+
+    # TODO: Should this be defined per-company?
```
Yes it should, Sylvain will use this in a multi-company context. To do so, use a related field like you did in  for `customer_wallet_account_id`.

robinkeunen commented on account_customer_wallet/models/res_config_settings.py:
```diff
@@ -0,0 +1,43 @@
+        # TODO: Filter domain on type?
+        domain="[('company_id', '=', company_id)]",
+        string="Customer Wallet Account",
+        config_parameter="account_customer_wallet.customer_wallet_account_id",
```
This is not needed - from what I understand - since you already write on res.company through the related field.

If you use config_parameter, res.config.settings writes in table `ir_config_parameter` which is NOT company specific.

robinkeunen commented on account_customer_wallet/models/res_config_settings.py:
```diff
@@ -0,0 +1,43 @@
+        help="The account where all wallet transactions will be recorded",
+    )
+
+    @api.onchange("customer_wallet", "customer_wallet_account_id")
+    def _onchange_customer_wallet(self):
+        """Do not allow *customer_wallet* to be enabled without defining
+        *customer_wallet_account_id*. There is probably a better way to do this.
+        """
+        res = {}
+        if self.customer_wallet and not self.customer_wallet_account_id:
+            self.customer_wallet = False
+            res["warning"] = {
+                "title": _("Error!"),
+                "message": _(
+                    "You cannot enable this setting without defining a customer"
+                    " wallet account."
+                ),
+            }
+        return res
```
Setting a required domain in the view is probably cleaner and easier. Something like

```xml
<field name="customer_wallet_account_id" attrs="{'required': [('customer_wallet', '=', True)]}"/>
```

carmenbianca replied:
This sadly doesn't work.

```xml
<field name="customer_wallet" attrs="{'required': [('customer_wallet_account_id', '!=', None)]}" />
```

Not sure why.

carmenbianca replied:
No longer needed.

robinkeunen commented on account_customer_wallet/models/res_partner.py:
```diff
@@ -0,0 +1,55 @@
+    def _compute_customer_wallet_account_id(self):
+        account_id = self.env["ir.config_parameter"].get_param(
+            "account_customer_wallet.customer_wallet_account_id"
+        )
```
Rather get it on the company.

robinkeunen commented on account_customer_wallet/models/res_partner.py:
```diff
@@ -0,0 +1,55 @@
+            "account_customer_wallet.customer_wallet_account_id"
+        )
+        result = self.env["account.account"].browse(int(account_id))
+        for partner in self:
+            partner.customer_wallet_account_id = result
```
When you set the same value on all records, you can use a write:

```py
self.write({"customer_wallet_account_id": customer_wallet_account_id})
```

carmenbianca replied:
Doesn't work now that we need to use each partner's company_id.

legalsylvain replied:
that case, maybe use a ``with_context(force_company=company_id)``

robinkeunen commented on account_customer_wallet/models/res_partner.py:
```diff
@@ -0,0 +1,55 @@
+class Partner(models.Model):
+    _inherit = "res.partner"
+
+    # TODO: What if this setting is changed during runtime? Does res.partner
```
It will if you rather fetch the setting on the company and add `@api.depends("company_id.customer_wallet_account_id")` to you compute function.

robinkeunen commented on account_customer_wallet/models/res_partner.py:
```diff
@@ -0,0 +1,55 @@
+    @api.depends("customer_wallet_account_id.move_line_ids")
+    def _compute_customer_wallet_balance(self):
+        for partner in self:
+            # TODO: Use filtered or search?
```
I'd go for a search since `customer_wallet_account_id.move_line_ids` could be a big bunch of line. Rather only fetch the needed lines with a search.

I'm not sure how deep the CRM pre-fetches the related fields.

carmenbianca replied:
No longer relevant.

robinkeunen commented on account_customer_wallet/models/res_partner.py:
```diff
@@ -0,0 +1,55 @@
+            #     ("partner_id", "=", partner.id),
+            # ])
+
+            # TODO: Is this correct?
```
who knows ?

robinkeunen commented on account_customer_wallet/views/res_config_settings_views.xml:
```diff
@@ -0,0 +1,36 @@
+        <field name="inherit_id" ref="account.res_config_settings_view_form" />
+        <field name="arch" type="xml">
+            <!--
+                The o_settings_container div annoyingly doesn't have an id of its own.
+                Using one of its children.
```
¯\_(ツ)_/¯

carmenbianca commented:
Copying some internal conversations over.

---

@polchampion said:

> [The above is] one of the highest-priority improvements that we could make if there is time next week, but we would need to talk about this some more to think about it. Notably, would the different pos sessions talk with each other [using the JS method] ?

---

@carmenbianca said:

> If implemented in JavaScript: the sessions will not talk with each other.
>
> I could instead attempt to implement this in Python. We currently have customer_wallet_balance as a computed field on res.partner. This is good and we'll keep that field as-is.
>
> We can then create another field pos_session_customer_wallet_balance that takes the value of customer_wallet_balance and adds/subtracts to it the unfinished lines from all active POS sessions.
>
> I'd have to check with @robinkeunen  to see if that is possible. We'd need to @api.depends on active POS session lines, but I'm not sure how to do that.

---

@polchampion said:

> According to @legalsylvain, I seem to recall that the  biggest challenge was to synchronise the POS while it's running, and the only easy way to do this is via a manual "refresh"-like button that the cashier would need to click on each open pos session.
>
> Is there a significant technical difference between these two options :
>
> - adding the computation of unfinished pos lines to the regular customer_wallet_balance field, and triggering its computation from the pos, when fetching the value of pos_sessions_customer_wallet_balance
> - keeping customer_wallet_balance as is and appending the unfinished pos lines to pos_sessions_customer_wallet_balance when fetching the value of the former ?

---

@carmenbianca said:

> I wouldn't do either option, or maybe there's a slight breakdown in communication.
>
> To answer your actual question on the difference between the two: Option 1 would introduce a hard dependency on `pos` and leave you with no way to compute the _verified_ balance. Option 2 I do not really understand.
>
> My proposal is effectively this: `customer_wallet_balance` works as-is. It gives you the balance of all verified `account.move.line`s. There is value in keeping that behaviour. This value will be displayed on the partner view form in Odoo.
>
> I would then add a second field `pos_session_customer_wallet_balance`. This is equal to `customer_wallet_balance - all payments made by the partner across all active POS sessions`. This field will be used and displayed inside of the POS.
>
> I think this can be done without a refresh button. Maybe.

---

@polchampion said:

> Ok, I think I understand. I can thing of two questions :
>
> 1. Is there a reason you would only display pos_session_customer_wallet inside of the pos sessions and not on the res.partner ?
> 2. And what would trigger this field's computation ?

---

@carmenbianca said:

> 1. We could make `pos_customer_wallet` cause the field that is displayed in res.partner to be replaced with `pos_session_customer_wallet_balance`. That's fine with me.
> 2. Good question :) I wrote about that [here](https://github.com/coopiteasy/addons/pull/222#issuecomment-1099983778) because I'm not sure.

robinkeunen reviewed (commented):
A first bunch of comments, good start. I did not review the js because I'm 💩  in js.

legalsylvain commented:
Hi thanks for the ping ! 

huge code, I didn't review all the points, but regarding that question : 


>    Write some JavaScript to calculate the current balance: partner.customer_wallet_balance - all payments made by the partner in the current session. Problem: Doesn't work across sessions.
    Create another field pos_session_customer_wallet_balance on partner that is equal to partner.customer_wallet_balance - all payments made by the partner across all active POS sessions. Problem: I'm not sure what @api.depends to use here. I think I'd have to create a computed Many2many field on res.partner that returns all active POS sessions, and I can then do something like @api.depends("pos_session_ids, pos_session_ids.payment_lines") (or whatever the actual field would be).

> @robinkeunen @legalsylvain What are your thoughts?

**option 1** : is not good I think. (because as you said, does'nt work over sessions + doesn't work inside a session I think if session is reloaded + javascript dev is a mess and should rewritten in V16, so, better to do it in python)
**option 2** : I was more thinking about overloading the current field _compute_customer_wallet_balance in the pos_customer_wallet module, calling super + adding pending account bank statement line.
Pro : that avoid to create an other field, and to have two field to handle in the algorithm to check if the wallet is debitable.
Doesn't solve : the problem regarding the @api.depends 

I think that we could make a dependency like that : 
in ``pos_customer_wallet/models/res_partner.py``

```
@depends("customer_wallet_account_id.move_line_ids", "statement_line_ids.amount")
 def _compute_customer_wallet_balance(self):
    super()._compute_customer_wallet_balance()
    for partner in self:
        partner.customer_wallet_balance += # get here_statement line from open sessions
```
Note sure it work 100%.


Requires to add in res_partner, the field ``statement_line_ids = fields.One2many(comodel_name="account.bank.statement.line", inverse_name="partner_id")``

Not sure if I'm clear. ;-)

carmenbianca committed:
[FIX] account_customer_wallet: Write account_wallet to res.company instead of settings

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] account_customer_wallet: Use search instead of filter

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] account_customer_wallet: Mark failing test as expectedFailure for now

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] pos_customer_wallet: Recompute balance based on POS statement lines

This does not yet live-update in the POS. The window needs to be
reloaded to see the updated balance.

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[REF] account_customer_wallet: Rename account.py to account_account.py

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca commented:
@legalsylvain This is your proposed solution: https://github.com/coopiteasy/addons/pull/222/commits/96ceb1f32a3ce44981354d115eb5425c7dc87db8

It works, BUT… The POS session does not fetch the recomputed value. The POS appears to cache the first value for `customer_wallet_balance`, and then never fetches any newer value. If I reload the browser window, the newest balance is shown.

There is probably an easy way to force-fetch the newest value in JavaScript, but I'll have to look into that later.

legalsylvain commented:
Hi @carmenbianca. thanks for the implementation.

> and then never fetches any newer value

Not sure to understand that point. In V12, PoS works fully offline, and doesn't reload product / partner data. so it is an awaited behaviour to have to A) reload the PoS or B) call a read when selecting a customer in the PoS.

what do you think ? 

(maybe I misunderstood something...)

carmenbianca commented:
@legalsylvain The problem that is [described here](https://github.com/coopiteasy/addons/pull/222#issuecomment-1099983778) is that the POS doesn't display the new wallet balance after the customer makes a payment in the POS. So if the customer comes back a second time, it still shows the balance as it was at the start of the POS session.

I may be a little confused, though. I don't know to what degree the POS is offline, because it does appear to write to the database.

If there's no way to fetch new data from inside the POS session (apart from reloading the window), then I think the most sensible thing to do is to update the cached value of `customer_wallet_balance` inside JavaScript.

legalsylvain commented:
Hi carmen.

>  I don't know to what degree the POS is offline, because it does appear to write to the database.

the PoS works offline means that you can make sales all the day without connexion, **once the PoS is loaded**. You can test it very easily : run an odoo locally, open a PoS via the url : http://localhost:xxxx/.../pos and then stop the server. you can so make sales. Once validated, you'll see that sales are stored in the local storage, and there is a number of sales at the right top of the screen. PoS will try to send sales when a new sale is done, or if cashier click on the number...

So : 
- Things to do : after a sale is confirmed in the PoS, update in javascript the value of the wallet of the customer you loaded at the beginning. So it will works offline / online. (in mono PoS configuration) and fix [this use case](https://github.com/coopiteasy/addons/pull/222#issuecomment-1099983778).

- Nice to have : when selecting a customer, refresh the data by calling a read in javascript. (that could of course fail, in offline situation). That way it will works online in multi PoS configuration.

Off course the "offline + MultiPoS" will never work, by design.

I can develop the second nice to have point, if it's not in the Coop It Easy roadmap. 

Let me know if I'm not clear.

carmenbianca commented:
Thank you so much for the detailed response @legalsylvain ! That makes perfect sense.

Pinging @polchampion to help make a call on the two points. I think I can do both quite comfortably.

polchampion commented:
> Pinging @polchampion to help make a call on the two points. I think I can do both quite comfortably.

@carmenbianca Let's go for the first thing. then if you have spent no more than 4-5 hours on this, it might be possible to develop the nice to have (refresh when selecting customer), depending on the additional time you will spend on the tests feedback, which will take priority. If in doubt, you can make a call.

legalsylvain commented:
Note on the second point, I think we could create a very simple ``pos_refresh_customer`` against ``OCA/pos`` that makes only that point. that's a very generic feature in fact that could be helpfull for other tools. 
What do you think ?
(no need to make an extra dependencies BTW. just that install the ``pos_refresh_customer`` will add a new feature). 

the refresh can be done by a button manually, or maybe automatically, when we select a customer. (I think there is a hook like ``_select_customer()`` ....

carmenbianca committed:
[ADD] pos_account_wallet: Update wallet balance in POS

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca commented:
@polchampion Good news—that took all of 10 minutes to implement.

carmenbianca commented:
@legalsylvain Pour info, j'essaierai de créer `pos_refresh_customer`.

carmenbianca committed:
[FIX] pos_customer_wallet: Update write_date on computation of wallet balance

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

legalsylvain commented on account_customer_wallet/models/res_company.py:
```diff
@@ -0,0 +1,18 @@
+    customer_wallet_account_id = fields.Many2one(
+        comodel_name="account.account",
+        string="Customer Wallet Account",
+        # TODO: Filter domain on type?
```
I don't think so. i think this will depends on localization or point of views of accountant. So let free is the better / simplier option.

legalsylvain commented on account_customer_wallet/models/res_config_settings.py:
```diff
@@ -0,0 +1,41 @@
+class ResConfigSettings(models.TransientModel):
+    _inherit = "res.config.settings"
+
+    # TODO: Should this be defined per-company?
```
comment could be removed so.

legalsylvain commented on account_customer_wallet/models/res_partner.py:
```diff
@@ -0,0 +1,52 @@
+        compute="_compute_customer_wallet_account_id",
+    )
+    # TODO: Is Monetary the correct field? What is behaviour with multiple
+    # currencies?
```
comment could be removed.

legalsylvain commented on account_customer_wallet/models/res_partner.py:
```diff
@@ -0,0 +1,52 @@
+    # currencies?
+    customer_wallet_balance = fields.Monetary(
+        string="Customer Wallet Balance",
+        currency_field="company_currency_id",
```
no need to set this field. and to defined ``company_currency_id``.

we can define this field the same way odoo does : 
https://github.com/odoo/odoo/blob/12.0/addons/account/models/partner.py#L389

legalsylvain commented on account_customer_wallet/models/res_partner.py:
```diff
@@ -0,0 +1,52 @@
+        compute="_compute_customer_wallet_balance",
+        readonly=True,
+    )
+    company_currency_id = fields.Many2one(
```
currency is available here on ``res.partner`` model.
that definition could be removed : 
https://github.com/odoo/odoo/blob/12.0/addons/account/models/partner.py#L391

legalsylvain commented on account_customer_wallet/models/res_partner.py:
```diff
@@ -0,0 +1,52 @@
+        "customer_wallet_account_id.move_line_ids",
+        "customer_wallet_account_id.move_line_ids.balance",
+    )
+    def _compute_customer_wallet_balance(self):
```
I think we could reuse same algorithm as in the ``_invoice_total`` function : 
https://github.com/odoo/odoo/blob/12.0/addons/account/models/partner.py#L264

To : 
- handle multicurrency. (I think that this point is optional, and can assume in the roadmap that the module works only in mono currency, but i think that one of your customer works in multi currency in switzerland...)
- handle parent / child computation.

carmenbianca replied:
I don't think the linked algorithm handles multicurrency, unless I'm mistaken and the database does some conversions.

legalsylvain replied:
you're right. That case, I think we should assume that the wallet is allways expressed in the company currency, and make conversion, if required. (but we can only raise an unimplementedError, if it's not in the scope of this PR).

legalsylvain commented on account_customer_wallet/models/account_account.py:
```diff
@@ -0,0 +1,16 @@
+
+    # TODO: Figure out whether there already exists a module that makes this
+    # One2many field available
+    move_line_ids = fields.One2many(
```
as ``_compute_customer_wallet_balance`` is stored to False, I think that we can remove the @api.depends in the funciton ``_compute_customer_wallet_balance``. Odoo core account does the same for the field ``total_invoiced``. 
https://github.com/odoo/odoo/blob/12.0/addons/account/models/partner.py#L389

It simplify the algorigthm and make the field ``move_line_ids`` unncessary.

carmenbianca replied:
@legalsylvain For clarity's sake: a compute function being `stored=False` means that the value will be recomputed every time the field is accessed? I was under the impression that the result is cached internally.

legalsylvain replied:
Well i just did the following : 
- edit /odoo/addons/account/models/partner.py and add a line ``print("Hello!")`` at the beginning of the function we want to copy ``def _invoice_total(self)`` that is computed / store = False.
- go the partner form.
- refresh your browser. each time you refresh, the print is displayed.

I so think the design is valid in that case to remove the @api.depends, and so simplify the code. (and also in the inherited function)

AFAIK, the only interest to make a depends on a non store fields, is to allow to recompute on the fly when depending fields are changing on the form view. 
if you have the folllowing  

```
field_A = fields.char()
field_B = fields.char(_compute=_compute_field_B, store=False)

def _compute_field_B(self):
```

if you add ``@api.depends("fields_A")``, in the form view, when you'll change the field A, the field B will be recomputed and change even if the form view is not saved. otherwise, the recompute will be fired when saving.

In our case, that feature is not interesting. Don't you thinK ?

carmenbianca replied:
Makes sense, I agree. I've implemented this in [0fe9b1a](https://github.com/coopiteasy/addons/pull/222/commits/0fe9b1a8841e4a92776ed46c09ea97de20ccd23e) and [a2ead73](https://github.com/coopiteasy/addons/pull/222/commits/a2ead7390b9479a396abf469ffa6e354375560b4).

legalsylvain reviewed (commented):
some extra remarks inline, regarding code simplification. 

If it's not clear, let me know, we can plan a little talk.

regards, and thank !

carmenbianca committed:
[FIX] account_customer_wallet: Remove some TODO comments

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[REF] account_customer_wallet: Use a related instead of a computed field

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[REM] account_customer_wallet: Remove some fields and api.depends lines

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

legalsylvain commented on account_customer_wallet/models/res_company.py:
```diff
@@ -0,0 +1,17 @@
+class Company(models.Model):
+    _inherit = "res.company"
+
+    customer_wallet = fields.Boolean(
```
I don't remember when I commented that point, but I think this field is unnecessary. we could just remove it and assume that if ``customer_wallet_account_id`` is defined, the feature is active. So it will simplify a lot the ``res.config.settings`` (no need to onchange and raise error). 
the ``is_enabled_customer_wallet`` of ``pos.config`` could became a computed non stored field, based on the ``company_id.customer_wallet_account_id`` field)

What do you think ?

carmenbianca replied:
Forwarding question to @polchampion

polchampion replied:
@carmenbianca I agree with @legalsylvain

carmenbianca replied:
:thumbsup: I remove some code then. I love removing code :D

legalsylvain replied:
;-)

carmenbianca replied:
[7974b1f](https://github.com/coopiteasy/addons/pull/222/commits/7974b1fd773e670140e6b639a60958baf6d379ae) code removed

carmenbianca committed:
[REF] account_customer_wallet: Re-implement wallet computation using partner._invoice_total algorithm

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] pos_customer_wallet: Tests

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[REF] pos_customer_wallet: Re-implement wallet computation using partner._invoice_total algorithm

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

legalsylvain commented:
nice refactor in ``tests/common.py`` !

legalsylvain committed:
[IMP] fully customize demo data on res_company, making the module ready to test, once installable with demo data

carmenbianca committed:
[REM] account_customer_wallet: Custom wallet toggle; depend on whether account is set instead

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

legalsylvain committed:
[ADD] test to see if balance is commown between child and parent partner

legalsylvain commented:
I broke something...

carmenbianca commented:
@legalsylvain I've been debugging this, because your PR definitely worked before it didn't anymore. I think _I_ broke something. `customer_wallet_balance` is not being reliably recomputed. I think I'll reintroduce the api.depends lines.

carmenbianca commented:
@legalsylvain So that didn't do anything :/ I'm force-removing the commits related to #228 and we'll try again in another branch.

carmenbianca committed:
[FIX] account_customer_wallet: Search balance on all partners in family tree

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] pos_customer_wallet: Avoid null error when no customer is selected

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] account_customer_wallet: README.rst

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
